### PR TITLE
Regions/neg adj3

### PR DESCRIPTION
**Pull request by oelbert**
## Purpose

Refactoring neg_adj3 to improve performance by removing Python code between stencil calls and by adding 2D fields 

## Code changes:

- `fillq` now uses 2d fields to accumulate sums
- Renamed internal stencil variables for clarity and consistency with the rest of the code
- Pulled Python code into stencils to allow full fusion
- Added a docstring to neg_adj3.compute()

### CODE REVIEW

File: fv3core/stencils/neg_adj3.py [hunk 1]
```diff
@@ -1,177 +1,151 @@
...
     pt: sd,
     dp: sd,
     delz: sd,
-    qv: sd,
-    ql: sd,
-    qr: sd,
-    qs: sd,
-    qi: sd,
-    qg: sd,
+    qvapor: sd,
```
Comment (eddie-c-davis) [inline fv3core/stencils/neg_adj3.py:123]: All occurrences of `sd` can now be replaced with `FloatField`.

File: fv3core/stencils/neg_adj3.py [hunk 2]
```diff
@@ -1,177 +1,151 @@
...
     with computation(BACKWARD), interval(...):
-        if q < 0. and sum1 >= 0:
+        if q < 0.0 and sum1 >= 0:
             dq = sum1 if sum1 < -q * dp else -q * dp
             sum1 = sum1 - dq
             sum2 = sum2 + dq
             q = q + dq / dp
     with computation(BACKWARD), interval(...):
-        if q > 0. and sum1 >= 1e-12 and sum2 > 0:
+        if q > 0.0 and sum1 >= 1e-12 and sum2 > 0:
```
Comment (eddie-c-davis) [inline fv3core/stencils/neg_adj3.py:112]: Nice touch, `0.` always looks incomplete to me...

File: fv3core/stencils/neg_adj3.py [hunk 3]
```diff
@@ -1,177 +1,151 @@
...
-def fillq(q:sd, dp:sd):
+def fillq(q: FloatField, dp: FloatField, sum1: FloatFieldIJ, sum2: FloatFieldIJ):
+    with computation(FORWARD), interval(...):
+        # reset accumulating fields
+        sum1 = 0.0
+        sum2 = 0.0
     with computation(FORWARD), interval(...):
         if q > 0:
-            sum1 = sum1[0, 0, -1] + q * dp
+            sum1 = sum1 + q * dp
```
Comment (eddie-c-davis) [inline fv3core/stencils/neg_adj3.py:104]: Love that this works now...

File: fv3core/stencils/neg_adj3.py [hunk 4]
```diff
@@ -1,177 +1,151 @@
...
-                    sum2 -= dq
-                    q[i, j, k] -= dq / dp[i, j, k]
-
-
-"""
-# TODO fix this to do fillq with a stencil that validates
-# need sum1, sum2 to be an accumulating floats
 @gtstencil()
-def fillq(q:sd, dp:sd):
+def fillq(q: FloatField, dp: FloatField, sum1: FloatFieldIJ, sum2: FloatFieldIJ):
```
Comment (eddie-c-davis) [inline fv3core/stencils/neg_adj3.py:97]: So awesome you were able to stencil-ize this.

File: fv3core/stencils/neg_adj3.py [hunk 5]
```diff
@@ -1,177 +1,151 @@
...
 from gt4py.gtscript import BACKWARD, FORWARD, PARALLEL, computation, interval
 
 import fv3core._config as spec
 import fv3core.utils.global_constants as constants
 import fv3core.utils.gt4py_utils as utils
 from fv3core.decorators import gtstencil
+from fv3core.utils.typing import FloatField, FloatFieldIJ
 
 
 sd = utils.sd
```
Comment (eddie-c-davis) [inline fv3core/stencils/neg_adj3.py:11]: `sd` can be removed in favor of `FloatField`.

File: fv3core/stencils/neg_adj3.py [hunk 6]
```diff
@@ -1,177 +1,151 @@
...
 ZVIR = constants.RVGAS / constants.RDGAS - 1.0
 
 
 @gtscript.function
-def fix_negative_ice(qv, qi, qs, qg, qr, ql, pt, lcpk, icpk, dq):
-    qsum = qi + qs
+def fix_negative_ice(qvapor, qice, qsnow, qgraupel, qrain, qliquid, pt, lcpk, icpk, dq):
```
Comment (eddie-c-davis) [inline fv3core/stencils/neg_adj3.py:16]: Type annotations are nice on `gtscript.function`'s too, but not required.

File: fv3core/stencils/neg_adj3.py [hunk 7]
```diff
@@ -206,99 +180,131 @@ def fix_neg_cloud(dp: sd, qcld: sd):
 
 
 # Nonstencil code for reference:
-def fix_water_vapor_nonstencil(grid, qv, dp):
+def fix_water_vapor_nonstencil(grid, qvapor, dp):
```
Comment (eddie-c-davis) [inline fv3core/stencils/neg_adj3.py:183]: Should we comment the non-stencil code?

File: fv3core/stencils/neg_adj3.py [hunk 8]
```diff
@@ -206,99 +180,131 @@ def fix_neg_cloud(dp: sd, qcld: sd):
...
         else:
             upper_fix = upper_fix[0, 0, 1]
+    with computation(FORWARD), interval(1, None):
+        upper_fix = upper_fix[0, 0, -1]
+    with computation(PARALLEL), interval(-1, None):
+        qvapor[0, 0, 0] = upper_fix[0, 0, 0]
 
 
 def compute(qvapor, qliquid, qrain, qsnow, qice, qgraupel, qcld, pt, delp, delz, peln):
+    """
```
Comment (eddie-c-davis) [inline fv3core/stencils/neg_adj3.py:283]: Docstrings, well done!

File: fv3core/stencils/neg_adj3.py [hunk 9]
```diff
@@ -322,28 +328,32 @@ def compute(qvapor, qliquid, qrain, qsnow, qice, qgraupel, qcld, pt, delp, delz,
...
     fix_water_vapor_down(
         qvapor,
         delp,
         upper_fix,
         lower_fix,
-        dp_bot,
         origin=grid.compute_origin(),
         domain=grid.domain_shape_compute(),
     )
-    qvapor[:, :, grid.npz] = upper_fix[:, :, 0]
```
Comment (eddie-c-davis) [inline fv3core/stencils/neg_adj3.py:346]: Way to remove that numpy code!

**Review by eddie-c-davis — CHANGES_REQUESTED** [review]
Excellent work! Just a couple of suggestions related to the typing.

File: fv3core/stencils/neg_adj3.py (hunk 7 above)
Comment (oelbert) [reply fv3core/stencils/neg_adj3.py:183]: Oops! I'd also be ok with just removing it at this point, though

File: fv3core/stencils/neg_adj3.py (hunk 6 above)
Comment (oelbert) [reply fv3core/stencils/neg_adj3.py:16]: Would you object to me not including them this PR? I think we'll be revisiting this file a bit in the future

Comment (eddie-c-davis) [reply fv3core/stencils/neg_adj3.py:15]: No objection at all!

File: fv3core/stencils/neg_adj3.py (hunk 7 above)
Comment (eddie-c-davis) [reply fv3core/stencils/neg_adj3.py:182]: Sounds good, we can save that for the next round too :)

**Review by eddie-c-davis — APPROVED** [review]
Looks great!

File: fv3core/stencils/neg_adj3.py [hunk 10]
```diff
@@ -1,184 +1,157 @@
...
-            qs = qsum
-        elif qs < 0.0:
-            qs = 0.0
-            qi = qsum
+        if qice < 0.0:
+            qice = 0.0
+            qsnow = qsum
+        elif qsnow < 0.0:
+            qsnow = 0.0
+            qice = qsum
```
Comment (jdahm) [inline fv3core/stencils/neg_adj3.py:23]: Great work on the clearer variable names!

**Comment by oelbert** [comment]
Purpose
Refactoring neg_adj3 to improve performance by removing Python code between stencil calls and by adding 2D fields

Code changes:
- fillq now uses 2d fields to accumulate sums
- Renamed internal stencil variables for clarity and consistency with the rest of the code
- Pulled Python code into stencils to allow full fusion
- Added a docstring to neg_adj3.compute()

**Pull request merged**